File: Baek-joon/1992/1992.cpp
#include <iostream>
#include <sstream>

using namespace std;

int matrix[64][64];
string compression(int xs, int xe, int ys, int ye) {
	if (xs  == xe && ys  == ye) {
		return to_string(matrix[xs][ys]);
	}
	int xmid = (xs + xe) / 2;
	int ymid = (ys + ye) / 2;

	string leftup = compression(xs, xmid, ys, ymid);
	string rightup = compression(xs, xmid, ymid+1, ye);
	string leftdown = compression(xmid+1, xe, ys, ymid);
	string rightdown = compression(xmid+1, xe, ymid+1, ye);

	if (leftup.length() == 1 && rightup.length() == 1 && leftdown.length() == 1 && rightdown.length() == 1) {
		if (leftup.compare(rightup) == 0 && leftdown.compare(rightdown) == 0 && leftup.compare(leftdown) == 0) {
			return leftup;
		}
		else {
			string str = "(";
			str.append(leftup);
			str.append(rightup);
			str.append(leftdown);
			str.append(rightdown);
			str.append(")");
			return str;
		}
	}
	else {
		string str = "(";
		str.append(leftup);
		str.append(rightup);
		str.append(leftdown);
		str.append(rightdown);
		str.append(")");
		return str;
	}
}

int main(void)
{
	int n;
	cin >> n;

	string str;

	for (int i = 0; i < n; i++) {
		cin >> str;
		int j = 0;
		for (string::iterator iter = str.begin(); iter != str.end(); ++iter) {
			matrix[i][j++] = *iter -48;
		}

	}
	//for (int i = 0; i < n; i++) {
	//	for (int j = 0; j < n; j++) {
	//		cout << matrix[i][j] << " ";
	//	}
	//	cout<<endl;
	//}
	cout << compression(0, n-1, 0, n-1)<<endl;

	cin >> n;
}

File: Algorithm_problem_solving/Baek-joon/1992/1992.kt
import java.io.File

fun main() {
    File("inputFile.txt").bufferedReader(Charsets.UTF_8).useLines { lines ->
        val iterator = lines.iterator()

        val n = iterator.next().trim().toInt()

        val matrix = Array(n) {
            val line = iterator.next()
            IntArray(n) { col -> line[col] - '0' }
        }

        val quadTree = QuadTree(n, matrix)
        print(quadTree.encode())
    }
}


class QuadTree(
    val n: Int, val matrix: Array<IntArray>
) {

    fun encode() = compression(0, 0, n)

    private fun compression(
        y: Int, x: Int, size: Int
    ): String {
        if (size == 1) return matrix[y][x].toString()
        val halfSize = size / 2;

        val leftUp = compression(
            y, x, halfSize
        );
        val leftDown = compression(
            y + halfSize, x, halfSize
        );
        val rightUp = compression(
            y, x + halfSize, halfSize
        );
        val rightDown = compression(
            y + halfSize, x + halfSize, halfSize
        );

        if (isUniform(leftUp, leftDown, rightUp, rightDown)) {
            return leftUp
        } else {
            return "($leftUp$rightUp$leftDown$rightDown)"
        }
    }

    private fun isUniform(
        leftUp: String,
        leftDown: String,
        rightUp: String,
        rightDown: String,
    ): Boolean {
        return leftUp.length == 1 && leftUp == leftDown && rightUp == rightDown && leftUp == rightUp
    }
}